File: C-language/basic/L03/hanoi.c
#include <stdio.h>

void hanoi(int n, char from, char to, char temp);
unsigned long contaMosse(int n);
//il numero minimo di mosse per risolvere il gioco delle torri di hanoi è di 2^n -1 dove n è il numero di dischi da spostare.
int main(){
    int num;
    unsigned long  mosse;
    char risp = '.';
    printf("Inserire il numero di dischi: ");
    scanf("%d",&num);
    printf("\n");
    printf("Vuoi vedere solo il numero di mosse necessarie? (s/n)\n");
    scanf(" %c",&risp);
    if(risp == 'n')
        hanoi(num,'0','2','1');
    mosse = contaMosse(num);
    printf("Numero mosse: %lu\n",mosse);
    return 0;
}

unsigned long contaMosse(int n){
    if(n == 0)
        return 0;
    return 2* contaMosse(n - 1) + 1;
}
void hanoi(int n,char from, char to, char temp){
    if(n == 1){
        printf("Sposto 1 da paletto %c a paletto %c\n",from,to);
	return;
    }
    hanoi(n-1,from,temp,to);
    printf("Sposto %d da paletto %c a paletto %c\n",n,from,to);
    hanoi(n-1,temp,to,from);
}

File: Go-language/basic/L03_ricorsione/hanoi.go
package main

import "fmt"

func main() {
	var n int
	var mosse uint64
	var risp string
	fmt.Print("Inserire il numero di dischi: ")
	fmt.Scan(&n)
	fmt.Print("Vuoi solo il numero di mosse? s/n")
	fmt.Scan(&risp)
	if risp == "n"{
		hanoi(n,string('0'),string('2'),string('1'))
	}
	mosse = contaMosse(n)
	fmt.Println("Numero mosse: ",mosse)
}

func hanoi(n int,from,to,temp string) {
	if n == 1{
		fmt.Printf("Sposto 1 da paletto %s a paletto %s\n",from,to)
		return
	}
	hanoi(n-1,from,temp,to)
	fmt.Printf("Sposto %d da paletto %s a paletto %s\n",n,from,to)
	hanoi(n-1,temp,to,from)
}

func contaMosse(n int) uint64 {
	if n == 0{
		return 0
	}
	return 2 * contaMosse(n-1) +1
}
